src/mixer.rs

```diff
@@ -0,0 +1,428 @@
+    }
+}
+
+/// Wraps [snd_mixer_selem_id_t](http://www.alsa-project.org/alsa-doc/alsa-lib/group___simple_mixer.html)
+#[derive(Copy, Clone)]
+pub struct Elem<'a>{
+    handle: *mut alsa::snd_mixer_elem_t,
+    mixer: &'a Mixer
+}
+
+impl<'a> Elem<'a> {
+    /// Checks if this element is null by checking if the handle is null
+    pub fn is_null(&self) -> bool {
+        self.handle == 0 as *mut alsa::snd_mixer_elem_t
+    }
+}
+
+
+/// Iterator for all elements of mixer
+#[derive(Copy, Clone)]
+pub struct Iter<'a>{
+    last_handle: *mut alsa::snd_mixer_elem_t,
+    mixer: &'a Mixer
+}
+
+impl<'a> Iter<'a> {
+    pub fn new(m: &'a Mixer) -> Iter<'a> {
+        Iter {
+            last_handle: ptr::null_mut(),
+            mixer: m
+        }
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = Elem<'a>;
+
+    fn next(&mut self) -> Option<Elem<'a>> {
+        let elem = if self.last_handle.is_null() {
+            unsafe { alsa::snd_mixer_first_elem(self.mixer.0) }
+        } else {
+            unsafe { alsa::snd_mixer_elem_next(self.last_handle) }
+        };
+
+        if elem.is_null() {
+            None
+        } else {
+            self.last_handle = elem;
+            Some(Elem { handle: elem, mixer: self.mixer})
+        }
+    }
+
+}
+
+/// Wrapper for `snd_mixer_selem_id_t`, using array of hardcoded length
+// #[derive(Copy, Clone)]
+pub struct SelemId([u8; SELEM_ID_SIZE]);
+
+impl SelemId {
+    /// Creates a new SelemId` of hardcoded size SELEM_ID_SIZE.
+    /// This size is checked against `snd_mixer_selem_id_sizeof`
+    pub fn new(elem: Elem) -> SelemId {
+        // Create empty selem_id and fill from mixer
+        let sid = SelemId::empty();
+        unsafe { alsa::snd_mixer_selem_get_id(elem.handle, sid.as_ptr()) };
+        sid
+    }
+
+    /// Returns an empty (zeroed) SelemId. This id is not a useable id and need to be initialized
+    /// like `SelemId::new()` does
+    pub fn empty() -> SelemId {
+        assert!(unsafe { alsa::snd_mixer_selem_id_sizeof() } as usize <= SELEM_ID_SIZE);
+        // Create empty selem_id and fill from mixer
+        SelemId(unsafe { mem::zeroed() })
+    }
+
+    /// Convert SelemId into ``*mut *mut snd_mixer_selem_id_t` that the alsa call needs.
+    /// See [snd_mixer_selem_id_t](http://www.alsa-project.org/alsa-doc/alsa-lib/group___simple_mixer.html)
+    #[inline]
+    pub fn as_ptr(&self) -> *mut alsa::snd_mixer_selem_id_t {
+        self.0.as_ptr() as *const _ as *mut alsa::snd_mixer_selem_id_t
+    }
+}
+
+/// Wraps an Elem as a Selem
+// #[derive(Copy, Clone)]
+pub struct Selem<'a>(SelemId, Elem<'a>);
+
+impl<'a> Selem<'a> {
+    /// Creates a Selem by wrapping `elem`.
+    pub fn new(elem: Elem<'a>) -> Selem<'a> {
+        Selem(SelemId::new(elem), elem)
+    }
+
+    /// Creates a Selem by lookung for a specific selem by name given a mixer (of a card)
+    pub fn find_by_name(mixer: &'a Mixer, name: &str) -> Result<Selem<'a>> {
```

Comment:
What's wrong with having a find_selem function on Mixer instead? Since this is basically a wrapper around snd_mixer_find_selem